jax/_src/interpreters/mlir.py

```diff
@@ -886,7 +886,8 @@ def lower_jaxpr_to_module(
   platforms_with_donation = [p for p in platforms
                              if p in _platforms_with_donation]
   if platforms_with_donation:
-    if len(platforms_with_donation) != len(platforms):
+    if len(platforms_with_donation) != len(platforms) and (
```

Comment:
I am actually wondering if we need this check at all. The donation is not a lowering property, but a compilation property, and when we compile and execute the platform has been fixed.

@yashk2810 wdyt?

Reply:
That's true but input_output_aliases and xla_donated_args are embedded into stableHLO. We don't want to do that for a platform that doesn't support lowering right?

[https://source.corp.google.com/piper///depot/google3/third_party/py/jax/_src/interpreters/mlir.py;l=1289-1301;rcl=644732071 an](https://github.com/google/jax/blob/main/jax/_src/interpreters/mlir.py#L1289-L1301) and https://github.com/google/jax/blob/main/jax/_src/interpreters/mlir.py#L1284-L1287

Reply:
I have had a lot of complaints before where we silently don't donate because of some mismatch and that is very frustrating. So it's always better to error loudly in cases of donation.

Reply:
As far as I can tell, the only thing that is embedded is an MLIR argument attribute `tf.aliasing_output = 0` to the top-level function. However, when the export is compiled it is always an inner function, because it is always embedded in a caller. It seems that the `tf.aliasing_output` is ignored when it is not on the top-level function, and it makes sense because it is the caller that should decide whether it donates or not.

Reply:
But this affects not just export but `.lower` too and that can happen at the top level.

Reply:
Currently, if you use `.lower()` at the top-level and you use it with multi-platform lowering, you will not be able to compile the result, except if called from the `export.call` API. And if you use it with `export.call` the donation that was specified at export time is overriden by the donation specified at the `export.call` time. But if we eventually implement that support, this error will be helpful.

Reply:
I think we should keep the error. Silently overriding users wish instead of erroring for donation is bad